builtin/consul/config_sourcer.go

```diff
@@ -0,0 +1,410 @@
+	return cs.stop
+}
+
+func (cs *ConfigSourcer) read(ctx context.Context, log hclog.Logger, reqs []*component.ConfigRequest) ([]*pb.ConfigSource_Value, error) {
+	log.Trace("Reading KV data from Consul")
+	// Setup our lock
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
+	// Create our cache if this is our first time
+	if cs.cache == nil {
+		cs.cache = make(map[string]*cachedVal)
+	}
+
+	// Create our Consul API client if this is our first time
+	if cs.client == nil {
+		client, err := cs.config.toAPIClient()
+		if err != nil {
+			return nil, fmt.Errorf("Invalid Consul client configuration: %w", err)
+		}
+		cs.client = client.KV()
+	}
+
+	var results []*pb.ConfigSource_Value
+	for _, req := range reqs {
+		result := &pb.ConfigSource_Value{Name: req.Name}
+		results = append(results, result)
+
+		// Decode our configuration
+		kvReq := reqConfig{
+			// default to allowing stale queries
+			AllowStale: true,
+		}
+		if err := mapstructure.WeakDecode(req.Config, &kvReq); err != nil {
+			result.Result = &pb.ConfigSource_Value_Error{
+				Error: status.New(codes.Aborted, err.Error()).Proto(),
+			}
+
+			continue
+		}
+
+		opts := &api.QueryOptions{
+			Namespace:  kvReq.Namespace,
+			Partition:  kvReq.Partition,
+			Datacenter: kvReq.Datacenter,
+			AllowStale: kvReq.AllowStale,
+		}
+		reqLogger := log.With("key", kvReq.Key, "stale", kvReq.AllowStale)
+		if kvReq.Namespace != "" {
+			reqLogger = reqLogger.With("namespace", kvReq.Namespace)
+		}
+		if kvReq.Partition != "" {
+			reqLogger = reqLogger.With("partition", kvReq.Partition)
+		}
+		if kvReq.Datacenter != "" {
+			reqLogger = reqLogger.With("datacenter", kvReq.Partition)
+		}
+
+		cacheVal, ok := cs.cache[kvReq.cacheKey()]
+		if !ok {
+			reqLogger.Trace("querying Consul KV")
+			kvpair, meta, err := cs.client.Get(kvReq.Key, opts.WithContext(ctx))
+			if err != nil {
+				result.Result = &pb.ConfigSource_Value_Error{
+					Error: status.New(codes.Aborted, err.Error()).Proto(),
+				}
+
+				continue
+			}
+
+			refreshCtx, cancel := context.WithCancel(context.Background())
+			cacheVal = &cachedVal{
+				kvPair: kvpair,
+				cancel: cancel,
+				err:    nil,
+			}
+			cs.cache[kvReq.cacheKey()] = cacheVal
+			cs.startBlockingQuery(refreshCtx, log, kvReq, meta.LastIndex, opts)
+		}
+
+		if cacheVal.err != nil {
+			result.Result = &pb.ConfigSource_Value_Error{
+				Error: status.New(codes.Aborted, cacheVal.err.Error()).Proto(),
+			}
+		}
+
+		if cacheVal.kvPair == nil {
+			result.Result = &pb.ConfigSource_Value_Error{
+				Error: status.New(codes.NotFound, fmt.Sprintf("Configuration for key %s doesn't exist", kvReq.Key)).Proto(),
+			}
+
+			continue
+		}
+
+		result.Result = &pb.ConfigSource_Value_Value{
+			Value: string(cacheVal.kvPair.Value),
+		}
+	}
+
+	return results, nil
+}
+
+func (cs *ConfigSourcer) startBlockingQuery(ctx context.Context, logger hclog.Logger, kvReq reqConfig, lastIndex uint64, opts *api.QueryOptions) {
+	go func() {
```

Comment:
I think we should rename `startBlockingQuery` to something else, and perhaps a more idiomatic golang way of writing this would be to move the `go` goroutine call to the invocation, so instead of `cs.startBlockingQuery(refreshCtx, log, kvReq, meta.LastIndex, opts)` you have `go cs.startBlockingQuery(refreshCtx, log, kvReq, meta.LastIndex, opts)` and then remove the `go func(){}()` here. The reason being is that it's more intuitive to the reader that calling `startBlockingQuery` is supposed to be operating in a goroutine. 

For example, reading these lines:

```
cs.cache[kvReq.cacheKey()] = cacheVal
cs.startBlockingQuery(refreshCtx, log, kvReq, meta.LastIndex, opts)
```

a reader might think that `cs.startBlockingQuery()` in fact blocks here, and we wait until it returns, because after all the method name has `Blocking` in it... when in reality `startBlockingQuery()` doesn't block at all... it launches a goroutine and returns, regardless of whatever might block in that goroutine.

Reply:
I do realize that even in Effective Go there is an example of this style where a function or method immediately launches a goroutine, I just feel it's clearer to the reader to do `go functionName()`. If the method didn't have `Blocking` in the name I may not have even brought this up 🤷

Reply:
I agree, I think it makes more sense to start a go routine on this function rather than making the whole body of this function wrapped in a `go func()`.

Reply:
@catsby I'm updating this for the function to be called using a goroutine, rather than the entire contents of the function being in an anonymous function, called as a goroutine.

Also, I think that the nomenclature might be adding to the confusion. Consul has "[Blocking Queries](https://developer.hashicorp.com/consul/api-docs/features/blocking)", which are "used to wait for a potential change using long polling." I'll add information about this to the godoc for `startBlockingQuery`, which I may rename to `startConsulBlockingQuery`.

Reply:
@paladin-devops that's good context to know about Consul queries, thanks